[Video] Redesign of the "Add version" dialog of Manage versions

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Redesign of the "Add version" dialog of the Manage versions dialog, as discussed in Slack
* Offer similar movies from the library at first (same imdb id/tmdb id or same name and year)
* Buttons for alternative search methods
  * file picker to search in the file system videos that were not added to the library
  * selection from the whole library if the wanted movie was not in the similar movies shortlist

![image](https://github.com/xbmc/xbmc/assets/489377/c06553e2-22b4-4e26-9e9c-1ec61cf09087)

This is a select dialog. The diff may look large but this is mostly shuffling / copying of existing code, followed by some refactor (more left to do ideally...) See the intermediate commits for the step by step transformations. Validations of various strange situations kept as is, no more no less.

Left to do: 
* there already is a message "Browse files" in strings.po but I wasn't sure if the context was similar enough to reuse it.

not caused by the PR, existing wrong behavior: extras of a movie are lost when turning the movie into a version.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
User feedback and team consensus on the evolution of the button functionality.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Sqlite and Maria DB, movies that have similar movies in db or not.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
Easier to use versions functionality.

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP committed:
[Video] Split the version addition by file picker into its own function

CrystalP committed:
[Video] Add selection from similar movies of the movie to be added as version

CrystalP committed:
[Video] Add return codes to file picker selection

CrystalP committed:
[Video] Build the new dialog structure. Refresh parent dialog after adding a version

CrystalP committed:
[Video] Add selection from whole library of the movie to be added as version

thexai commented on xbmc/video/dialogs/GUIDialogVideoManagerVersions.cpp:
```diff
@@ -464,22 +423,17 @@ bool CGUIDialogVideoManagerVersions::ConvertVideoVersion(const std::shared_ptr<C
+
     if (list[i]->GetVideoInfoTag()->m_iDbId == dbId)
-    {
       list.Remove(i);
```
Remove items in same loop causes not all items processed as `i < list.Size()` condition changes dynamically: this is the cause full paths are not (always) displayed. Depends on if item to remove is before or after to item to keep.

I propose this or similar:

```c++
  for (int i = 0; i < list.Size(); ++i)
    list[i]->SetLabel2(list[i]->GetVideoInfoTag()->m_strFileNameAndPath);

  const int videoAssetDbId = m_videoAsset->GetVideoInfoTag()->m_iDbId;
  for (int i = 0; i < list.Size(); ++i)
  {
    if (list[i]->GetVideoInfoTag()->m_iDbId == videoAssetDbId)
    {
      list.Remove(i);
      break;
    }
  }
```

In this way full path is displayed always:

![Captura de pantalla 2023-12-30 102714](https://github.com/xbmc/xbmc/assets/58434170/871bbe06-644d-4a2c-9e44-9aa093c28115)

CrystalP replied:
Thanks, I didn't notice the side effect when I combined the loops.
Kind of ugly but works?
```
  for (int i = 0; i < list.Size(); ++i)
  {
    list[i]->SetLabel2(list[i]->GetVideoInfoTag()->m_strFileNameAndPath);

    if (list[i]->GetVideoInfoTag()->m_iDbId == videoAssetDbId)
    {
      list.Remove(i);
      --i;
    }
  }
```

ksooo replied:
I would go with:

```c++
  for (int i = 0; i < list.Size();)
  {
    list[i]->SetLabel2(list[i]->GetVideoInfoTag()->m_strFileNameAndPath);

    if (list[i]->GetVideoInfoTag()->m_iDbId == videoAssetDbId)
      list.Remove(i);
    else
      ++i;
  }
```

thexai replied:
> Kind of ugly but works?

Yes, this code is hard to understand. The basic concept is that when item is removed same `i` position is occupied with other item that needs to be processed (add path). Then first remove, second process same `i` position in same iteration:

Not tested but this "should" work:
 
```c++
  for (int i = 0; i < list.Size(); ++i)
  {
    if (list[i]->GetVideoInfoTag()->m_iDbId == videoAssetDbId)
      list.Remove(i);

    if (i < list.Size())
      list[i]->SetLabel2(list[i]->GetVideoInfoTag()->m_strFileNameAndPath);
  }
```

CrystalP replied:
Looks good to me. The single loop will make a difference for the Browse library with thousands of items.

ksooo replied:
What I suggested is well proved and done like tat in many other projects. IMO easy to read and understand.

ksooo replied:
The trick is to medial increase the counter.

CrystalP replied:
yeah it's a disguised while loop. I hesitated to rewrite it as such.

@thexai that suggestion doesn't work because it will still skip an item. That's why I added the --i in my ugly version.

thexai commented:
I think this PR is a very important improvement in usability. For the first time I have the feeling that the "Add version" button does what I expected. If the same change is implemented also in the context menu, it is almost perfect for me.

ksooo commented on xbmc/video/dialogs/GUIDialogVideoManagerVersions.h:
```diff
@@ -28,7 +28,19 @@ class CGUIDialogVideoManagerVersions : public CGUIDialogVideoManager
+   * whole library the movie that will be turned into a version.
+   * \return True for success and a version was created and attached, false otherwise.
+   */
+  static bool ConvertVideoVersion(const std::shared_ptr<CFileItem>& item, bool itemIsSource);
```
Pretty please, can we avoid bool parameters. The reader of the calling code "never" knows what that f***k true or false means at the first glance (without looking at doxy/header). Can we please introduce an enum class with two speaking values and get rid of the bool param?

CrystalP replied:
Working on it...

ksooo commented on xbmc/video/dialogs/GUIDialogVideoManagerVersions.h:
```diff
@@ -45,15 +57,53 @@ class CGUIDialogVideoManagerVersions : public CGUIDialogVideoManager
                                                   int dbId,
-                                                  CVideoDatabase& videoDb);
+                                                  CVideoDatabase& videoDb,
+                                                  bool itemIsSource);
```
No bool params, please.

ksooo reviewed (commented):
Very nice. Despite from not always working item-removing for loops, just some code style things I kindly ask to change.

ksooo commented:
>  If the same change is implemented also in the context menu

Sorry, we decided to remove the context menu item.

CrystalP commented on xbmc/video/dialogs/GUIDialogVideoManagerVersions.cpp:
```diff
@@ -531,5 +485,161 @@ bool CGUIDialogVideoManagerVersions::ProcessVideoVersion(VideoDbContentType item
+  return ChooseVideoAndConvertToVideoVersion(list, itemType, mediaType, dbId, videodb, true);
+}
+
+bool CGUIDialogVideoManagerVersions::AddVideoVersionFilePicker()
```
This here is not a duplicate, I moved all the existing code under that name in the previous commit.
Maybe for git blame purposes I should have left it in place and renamed the function, I'll see later if it doesn't cause massive rebase conflicts.

ksooo replied:
All good, then. Leave it as it is. Thanks.

CrystalP commented on xbmc/video/dialogs/GUIDialogVideoManagerVersions.cpp:
```diff
@@ -177,102 +182,53 @@ void CGUIDialogVideoManagerVersions::SetDefaultVideoVersion(const CFileItem& ver
 
-  CServiceBroker::GetMediaManager().GetLocalDrives(sources);
-  CServiceBroker::GetMediaManager().GetNetworkLocations(sources);
+  CGUIDialogSelect* dialog =
+      CServiceBroker::GetGUI()->GetWindowManager().GetWindow<CGUIDialogSelect>(
+          WINDOW_DIALOG_SELECT);
```
Taken care of here and in the other commits.

ksooo replied:
Thanks.

thexai commented:
![Captura de pantalla 2023-12-30 191530](https://github.com/xbmc/xbmc/assets/58434170/63e8a23a-c33d-4d2a-aa70-106e7e12de1d)


Minor cosmetic only: maybe move buttons order --> first "Browse library" as seems more used and logical order:

1- Short List
2- List off all movies in library
3- Any media file even is not in library

Hitcher commented:
This is much better, thanks.

EDIT: One problem still persists and that's the separate versions still appearing in the library list after managing them. Exiting the library and re-entering fixes it but a container refresh when closing the dialog would fix it.

CrystalP commented:
Reworked using previous feedback, changed the dialog title when displaying similar movies.

Q: when there are no similar movies, I'd like to display a disabled item in the list with the text "No similar movies in library", but I didn't find a property of the ListItem to do that, and disabling the select dialog blocks everything, including the Browse buttons that need to remain active. Ideas?

Noticed another issue: when picking the new version type, both versions and user extras types are displayed. Thought that was fixed by @ksooo earlier, I may have broken it, will check.

re: order of the buttons, I don't have a strong preference but thought that "Browse Library" would be for weird non-standard cases. Can be discussed by team in Slack, I'm sure there will be feedback on the forums as well about this...

All my testing was done from info > manage so far, will give the context menu a shot.

@Hitcher not sure what you're talking about and whether it's caused by my changes? Sounds like a separate issue to raise?

CrystalP commented on xbmc/video/dialogs/GUIDialogVideoManagerVersions.cpp:
```diff
@@ -531,5 +485,161 @@ bool CGUIDialogVideoManagerVersions::ProcessVideoVersion(VideoDbContentType item
+      list.Remove(i);
+  }
+  return true;
+}
```
fixed the same way as the other similar loop.

jjd-uk commented:
> Q: when there are no similar movies, I'd like to display a disabled item in the list with the text "No similar movies in library", but I didn't find a property of the ListItem to do that, and disabling the select dialog blocks everything, including the Browse buttons that need to remain active. Ideas?

What about something like this if it's not too much work

![image](https://github.com/xbmc/xbmc/assets/5781142/67faf914-4917-4357-968c-82b2500af0d2)

Where `<None>` is displayed with an empty source list.

thexai commented:
> changed the dialog title when displaying similar movies.

The title in full list dialog (Browse library button) still is "Add as version to..."

ksooo commented:
If you have no candidates for versions in your library, you are presented an empty list after pressing add version. This is somehow unexpected by end users, although "technically" correct. I think users will face this situation quite often, so not sure it is a good idea to show that empty list... 

<img width="1122" alt="image" src="https://github.com/xbmc/xbmc/assets/3226626/4d511a4f-8fc9-4dab-8e8f-71adf378b6e6">

ksooo commented:
> Noticed an issue: when picking the new version type, both versions and user extras types are displayed. Thought that was fixed by @ksooo earlier, I may have broken it, will check.

Cannot reproduce. I'm seeing only version types, no extra types. Hmm.

CrystalP commented:
re user extra names mixed with versions in the version type selection: was an issue in the data migration to v127. This PR is OK, I opened a dedicated issue.

> What about something like this if it's not too much work
> Where `<None>` is displayed with an empty source list.

That's the look (I figured out to do that), but not the feel. If you select <None>, it has the action of closing the dialog and opening a keyboard. I would like no action at all, with disabled look if possible.
Maybe there is a way to not auto-close the dialog, then the code could just "eat" the select event and do nothing.

> The title in full list dialog (Browse library button) still is "Add as version to..."

Changed.

> If you have no candidates for versions in your library, you are presented an empty list after pressing add version. This is somehow unexpected by end users, although "technically" correct. I think users will face this situation quite often, so not sure it is a good idea to show that empty list... Maybe in this case we should start with displaying either files or library browser,

Yes I was asking about that. I'm looking for a way to show a single disabled item in the list, with no action when selected.
Or is there a way to have the list hidden and replaced by a label?
The problem with opening files or library browser is that the user can't choose which one to open. Would need to popup a question first.

ksooo commented:
> Yes I was asking about that.

Oh, overlooked the question in the first place, sorry.

> I'm looking for a way to show a single disabled item in the list, with no action when selected.

This is something we try to avoid in our GUI. If there is something in an item list, I expect that I can do something with it.

> Or is there a way to have the list hidden and replaced by a label?

That would be the way to go and can be done easily by the skin, afaik.

Hitcher commented:
> @Hitcher not sure what you're talking about and whether it's caused by my changes? Sounds like a separate issue to raise?

Yeah, it was there before this. I'll raise a bug report.

CrystalP commented:
> > Or is there a way to have the list hidden and replaced by a label?
> 
> That would be the way to go and can be done easily by the skin, afaik.

@Hitcher @jjd-uk @jurialmunkey Is it possible? Does the code need to provide something to help?

Hitcher commented:
> Is it possible? Does the code need to provide something to help?

Just tested and a label with this visible condition works: `<visible>String.IsEmpty(Container(6).ListItem.Label)</visible>`

CrystalP commented:
It's good there is a solution. It feels hack-ish but good enough or not?
I don't know skinning but that label would appear for all select dialogs launched by manage versions and that have an empty list? Or this one specifically?

Would be better for the message to be translated, I could allocate a message in strings.po

Maybe there is a simple way to tweak the code of the select dialog, I'll look

jurialmunkey commented:
I think label in skin is best way to go. Will still work in existing skins without issue (just list appears blank) and doesn't force design decisions on the skin (skin can control location and appearance of label as appropriate to the skin design). 

The visibility conditions and localisation can be easily dealt with in a separate skin PR. Important thing is that the dialog set a unique window id/name for the skin to check against (see how ksooo did for select for versions/extras).

CrystalP commented:
Found a way to do the same thing through code so no skinning will be needed.
But maybe separate PR will be requested for the small change to GUIDialogSelect

I sort of understood the unique window id/name stuff, but that could take longer for me to finalize something.

jurialmunkey commented:
>  But maybe separate PR will be requested for the small change to GUIDialogSelect

I don't really understand this. You'd need a skin change anyway to accommodate whatever control you're planning to add. 

Why break existing skins by changing DialogSelect to make a label visible? It's not like the label is crucial to functionality. It seems limiting to have what is basically just hint text hard coded into the GUI via core rather than being a skin design choice.

> I sort of understood the unique window id/name stuff, but that could take longer for me to finalize something.

There's other approaches we can use to identify the dialog. Window ID is best in terms of options it gives on skin level, but simply setting a window property would also work for the purposes of a label visibility condition.

e.g. set window property "IsSimilarVersions" to True then skin can do:
`<visible>!String.IsEmpty(Window.Property(IsSimilarVersions)) + Integer.IsEqual(Container(6).NumItems,0) + !Container(6).IsUpdating</visible>`

CrystalP commented:
Give the lastest a try. I didn't add a label or other control and just tweaked the list a bit. The end result looks the same to me and is acceptable I think.

jurialmunkey commented:
Here's an example of some issues with hard coding a disabled listitem using current PR.

In my skin I use radiobutton style checkmark icons in DialogSelect lists to indicate the `IsSelected` flag because IMHO textcolour is a confusing signifier (does a different colour mean selected? unselected? focused? disabled?). Having a check on/off icon helps removes any ambiguity (checked is selected; unchecked is not selected).

The problem now is that the because the text is added as a listitem, it makes it look like I should be able to select/focus it (see ksooo's earlier point). This is made worse by Container.NumItems showing 1 item in list and not having any control over the style of the label to differentiate it from a listitem.

I know it is super nitpicky and I don't want to sound overly critical as overall PR in general is a huge improvement. But I do think it is important to avoid forcing skins into UX design corners unnecessarily.


![screenshot00341](https://github.com/xbmc/xbmc/assets/4618719/270e8adf-1a58-4476-8cd1-c46fdcf53f6b)

ksooo commented:
I fully agree to what @jurialmunkey said. This is something that should be handled by the skin. It is NOT a hack to do it like this.

Hitcher commented:
Couldn't we just change the heading label to display the text 'No similar movies found in the library.' instead?

Hitcher commented:
How did I end up closing this? I only replied to 2 posts!

ksooo commented:
> Couldn't we just change the heading label to display the text 'No similar movies found in the library.' instead?

Imo, this would be the worst ux I ever had, tbh. I really really would not like that.

Hitcher commented:
> I fully agree to what @jurialmunkey said. This is something that should be handled by the skin. It is NOT a hack to do it like this.

Yeah, let's stick to letting skins decide what they want to display when there's no items here. As said, they don't have to do anything and this will still function.

As an alternative to a label in the same place as the first list item and to avoid the confusion this may cause I propose a textbox with a message worded something like this:

![screenshot00004](https://github.com/xbmc/xbmc/assets/133808/ff0abe24-1ae9-4dd2-8318-d4c01253fb89)

Code used:

```
			<control type="textbox">
				<left>40</left>
				<top>120</top>
				<width>840</width>
				<height>588</height>
				<align>center</align>
				<aligny>center</aligny>
				<font>font14</font>
				<label>[B]No similar movies found.[/B][CR]Use the "Browse files" or "Browse library" buttons instead.</label>
				<visible>String.IsEmpty(Container(6).ListItem.Label)</visible>
			</control>
```

jjd-uk commented:
> As an alternative to a label in the same place as the first list item and to avoid the confusion this may cause I propose a textbox with a message worded something like this:

That looks good to me.

CrystalP commented:
OK I'll work on the window id and let you submit a skin PR.
How do translations work when you add messages with the skin? Do I need to do something?

I still think that core should support this situation a little better with a standard label for a information message that skinners can do whatever they want with. But that's not for today.

ksooo commented:
Don’t kill me, but can’t we in case we have no suggestions what to convert, just fill the list with either files or library items and omit the “other” button?

Hitcher commented:
> How do translations work when you add messages with the skin? Do I need to do something?

Personally I'd just add it after these ones:

https://github.com/xbmc/xbmc/commit/10c0811d30d5e521cc0f3509d7fb81f6f451c086#diff-71c048a41de97e0c50f5e628e8dfdf8d81f40b2e9133d5824150fe3663c01d79

ksooo commented:
> Personally I'd just add it after these ones:

Yes, adding strings to the global strings.po (instead skin-specific strings.po) is what we usually do with "skin only" strings (by means not used by the core) if we think they are of use for more than one skin.

CrystalP commented:
OK it will be 40032. What should I use as source code file name?

CrystalP committed:
[Video] Offer user to browse files or library when no similar movies can be found

CrystalP committed:
[Video] Refactor retrieval of similar and all movies

One step towards future extraction of that code from the dialog into a dedicated place.

Hitcher commented:
> What should I use as source code file name?
> 
> edit: The window id will be 12018 WINDOW_DIALOG_ADD_VIDEO_VERSION "addvideoversion" - or should it be more specific like WINDOW_DIALOG_ADD_SIMILAR_VIDEO_VERSION? window id to be used also on Browse library (currently WINDOW_DIALOG_SELECT) or just the similar movies?
> 
> Though it's not clear what happened to 12017 WINDOW_DIALOG_MANAGE_VIDEO_EXTRAS, which doesn't have a text in strings.po

I can't help with that, sorry.

ksooo commented:
Sorry, totally confused. Why should we need a new window id???

CrystalP commented:
@jurialmunkey requested one. It's one commit, easy to revert if overkill.

Put the last touches, the context menu removal gave the opportunity for more refactors / cleanups. There are still some left, but little by little the functions are split into components with more logical names, with the idea of extracting them from the dialog into a separate class

See my spin on the double-loop removal (consolidated across the dialog in one commit, before that it's just c/p from original code)

ksooo commented:
@jurialmunkey  for what do you need a new window is? I feel like we are starting to over engineer this thing…

Hitcher commented:
> for what do you need a new window is? I feel like we are starting to over engineer this thing…

I _think_ it was so skins know the context the dialog is being used for so they decide what label to use if the list is empty.

CrystalP commented:
Let's get this PR merged, I fixed the formatting for Jenkins, removed the window id commit and saved it so it could be easily re-added via other PR if needed.

jurialmunkey commented:
> @jurialmunkey for what do you need a new window is? I feel like we are starting to over engineer this thing…

@ksooo 
It is to differentiate it from other DialogSelect dialogs. The suggested "no items found" message text is very specific and makes no sense for other usage of the dialog. Needs to be a way to make it conditional on the type of select dialog. 

As I mentioned above, a simple window property would also suffice if window ID is too complicated.

Me personally I could take it or leave it. I'm not entirely convinced that a select dialog with no items should ever be shown (if there are no items, what are you selecting?). 

If it were me, in the case of no results for similar I'd instead pop-up a DialogConfirm with a message to say no results and three buttons: Library, Files, Cancel.

Hitcher commented:
I also don't think there's ever been a time when I've opened a select dialog and it was empty so it's probably exclusive to this scenario.

Hitcher commented:
@CrystalP my visible condition for the textbox/label needs updating.

`<visible>Control.IsVisible(6) + String.IsEmpty(Container(6).ListItem.Label)</visible>`

CrystalP commented:
I don't know what a window property is.
Didn't find a DialogConfirm, what's the exact name?
Since there are two options I could repurpose a DialogYesNo.

Hitcher commented:
`yesnodialog`

https://kodi.wiki/index.php?title=List_of_Built_In_Controls#DialogConfirm.xml

jurialmunkey commented:
> I also don't think there's ever been a time when I've opened a select dialog and it was empty so it's probably exclusive to this scenario.

Very easy for a python addon to open a blank DialogSelect because it didn't check that its list contained items first. In which case the message would make no sense:
```
import xbmcgui
mylist = [i for i in ['Mon', 'Wed', 'Fri'] if i.startswith('S')]
xbmcgui.Dialog().select('Weekends', mylist)
```

jurialmunkey commented:
> I don't know what a window property is. Didn't find a DialogConfirm, what's the exact name? Since there are two options I could repurpose a DialogYesNo.

@CrystalP

Not sure about in C++ but in Python the one with three buttons is `Dialog().yesnocustom` 
I think it is same ID as plain yesno but not 100% sure.
https://xbmc.github.io/docs.kodi.tv/master/kodi-base/d6/de8/group__python___dialog.html#ga82d498fed5e6282e3b061b084a8b574c

In skin yesno, yesnocustom and progress are all DialogConfirm.xml

CrystalP commented:
Yes thank you I found it and how to use it. The overloads on `CGUIDialogYesNo::ShowAndGetInput()` are a mess...
A refactor would cleanup a bit of redundancy but I don't have a good idea how at this time.

Let the bike shedding on the button placement begin... Though I'd prefer a followup PR after more feedback.

ksooo commented:
I can only repeat, that we should never show a select dialog with nothing to select. And to say the user that they should press one of the other buttons because there will be something to select then is just meh. Why not open either the file browser or the library browser. We can roll the dice to decide which one of these, but in any case better than the empty select list.

ksooo commented:
I just deleted my l last comment. In case you already read it, please ignore. I’m fine with the yesnodialog approach.

CrystalP commented:
Problem with just picking one (let's say Browse files for the argument) is that there will be no way for the user to go to Browse library.
(without significant changes to the code to offer buttons in Browse library and Browse files to switch from one to the other)

ksooo commented:
As I said, ignore my comment, which i meanwhile deleted.

CrystalP commented:
yeah I submitted exactly at the time your next message appeared.

Feel free to push the button if you're OK with the PR, I won't be around much longer today, party time!

jurialmunkey commented:
> Let the bike shedding on the button placement begin... Though I'd prefer a followup PR after more feedback.

Well no complaints from me since I suggested it.

> Problem with just picking one (let's say Browse files for the argument) is that there will be no way for the user to go to Browse library.

Other reason I suggested confirm was because I do think it is important to give the user *some* feedback about why the the usual dialog didn't open. Confirm seems the sensible way to provide a notification and give a choice on how to deal with it.

>  I won't be around much longer today, party time!

Enjoy! Well earned! 
I'm in the future, so I'm already nursing the 11am morning hangover...

Hitcher commented:
> Yes thank you I found it and how to use it. The overloads on `CGUIDialogYesNo::ShowAndGetInput()` are a mess... A refactor would cleanup a bit of redundancy but I don't have a good idea how at this time.
> 
> Let the bike shedding on the button placement begin... Though I'd prefer a followup PR after more feedback. ![image](https://private-user-images.githubusercontent.com/489377/293525831-c0222e42-6af5-4261-9a8a-4d2aae56d6f6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQwNzE5OTEsIm5iZiI6MTcwNDA3MTY5MSwicGF0aCI6Ii80ODkzNzcvMjkzNTI1ODMxLWMwMjIyZTQyLTZhZjUtNDI2MS05YThhLTRkMmFhZTU2ZDZmNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEwMVQwMTE0NTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lZWJhZGVmNGNlZWEzOWYxZDk5ODMwNTQ3ZWI4MDczNDkwMGU3MzEyNWJjZTA4NDEzYzVkYzEyMDYwNTlhYzdmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Das4QArK7K1JkULUbVe37xktb3rnxLQX6CebJL9kfrk)

Looks good to me and makes sense.

ksooo commented on xbmc/video/dialogs/GUIDialogVideoInfo.cpp:
```diff
@@ -53,6 +53,7 @@
 #include "video/VideoInfoTag.h"
 #include "video/VideoItemArtworkHandler.h"
 #include "video/VideoLibraryQueue.h"
+#include "video/VideoManagerTypes.h"
```
Is this new include really needed? You did not make any changes in this cpp file in this PR.

CrystalP replied:
left over from before the removal of context menu "Add as version to"

ksooo commented on xbmc/video/dialogs/GUIDialogVideoManagerVersions.cpp:
```diff
@@ -177,102 +182,99 @@ void CGUIDialogVideoManagerVersions::SetDefaultVideoVersion(const CFileItem& ver
 
-    const int idVideoVersion{m_database.GetVideoVersionInfo(path, idFile, typeVideoVersion, idMedia,
-                                                            itemMediaType, videoAssetType)};
+    const int dlgResult = CGUIDialogYesNo::ShowAndGetInput(
```
initializer vs. assignment

CrystalP replied:
Sure. ChatGPT, please rewrite my PR in the style of @ksooo :-)

ksooo commented on xbmc/video/dialogs/GUIDialogVideoManagerVersions.cpp:
```diff
@@ -177,102 +182,99 @@ void CGUIDialogVideoManagerVersions::SetDefaultVideoVersion(const CFileItem& ver
-        return;
-      }
+      case -1:
+      case 2:
+        // Cancel or Custom/Cancel
+        return false;
 
-      if (itemMediaType == MediaTypeMovie)
+      case 0:
```
I assume there are no defines / constants for these magic numbers? Just asking.

CrystalP replied:
No there weren't. It made things harder when implementing so I added some.
And used the existing NO_TIMEOUT.

ksooo commented on xbmc/video/dialogs/GUIDialogVideoManagerVersions.cpp:
```diff
@@ -177,102 +182,99 @@ void CGUIDialogVideoManagerVersions::SetDefaultVideoVersion(const CFileItem& ver
-              StringUtils::Format(g_localizeStrings.Get(40017), typeVideoVersion, videoTitle)))
-      {
-        return;
+        return ChooseVideoAndConvertToVideoVersion(
+            items, m_videoAsset->GetVideoContentType(), m_videoAsset->GetVideoInfoTag()->m_type,
+            m_videoAsset->GetVideoInfoTag()->m_iDbId, videoDb, MediaRole::Parent);
```
Real nitpicking: I'm always trying to avoid same function calls in a row if I can. Especially if I know that the implementation is not inline / not a one-liner. Looking at the implementation of `CFileItem::GetVideoInfoTag`, I think we should introduce a variable here instead calling it two times in a row.

```c++
CVideoInfoTag* CFileItem::GetVideoInfoTag()
{
  // Note: CPVRRecording is derived from CVideoInfoTag
  if (m_pvrRecordingInfoTag)
    return m_pvrRecordingInfoTag.get();
  else if (!m_videoInfoTag)
    m_videoInfoTag = new CVideoInfoTag;

  return m_videoInfoTag;
}

const CVideoInfoTag* CFileItem::GetVideoInfoTag() const
{
  // Note: CPVRRecording is derived from CVideoInfoTag
  return m_pvrRecordingInfoTag ? m_pvrRecordingInfoTag.get() : m_videoInfoTag;
}
```

CrystalP replied:
Agreed.

ksooo commented on xbmc/video/dialogs/GUIDialogVideoManagerVersions.cpp:
```diff
@@ -452,20 +507,177 @@ bool CGUIDialogVideoManagerVersions::ProcessVideoVersion(VideoDbContentType item
     {
-      list.Remove(i);
-      break;
+      CFileItemList versions;
+      m_database.GetVideoVersions(itemType, dbId, versions);
+      if (std::any_of(versions.begin(), versions.end(),
+                      [idFile](const std::shared_ptr<CFileItem>& version)
+                      { return version->GetVideoInfoTag()->m_iDbId == idFile; }))
```
Could you please add a `@todo` here. It seems we're obtaining all versions just to look for one specific item, then. As far as I can see, `versions` is not used elsewhere in this method.

CrystalP replied:
ok

ksooo commented on xbmc/video/dialogs/GUIDialogVideoManagerVersions.cpp:
```diff
@@ -452,20 +507,177 @@ bool CGUIDialogVideoManagerVersions::ProcessVideoVersion(VideoDbContentType item
+      continue;
+    }
+
+    list[i]->SetLabel2(tag->m_strFileNameAndPath);
```
`list[i]`is used twice in this loop. Looking at the implementation of `CFileItemList::operator[]`, a variable seems to be valuable here.

```c++
CFileItemPtr CFileItemList::operator[] (int iItem)
{
  return Get(iItem);
}

CFileItemPtr CFileItemList::Get(int iItem) const
{
  std::unique_lock<CCriticalSection> lock(m_lock);

  if (iItem > -1 && iItem < (int)m_items.size())
    return m_items[iItem];

  return CFileItemPtr();
}
``

CrystalP replied:
I didn't notice, ok.

ksooo reviewed (commented):
Only some nitpicks...

CrystalP committed:
[guilib] Add enum for results of CGUIDialogYesNo

CrystalP committed:
[Video] Add TODO for future improvement

ksooo reviewed (approved):
Looks really very good. Thanks for your patience with all the nitpicking end bike shedding. 👍

CrystalP commented:
The Windows UWP builds seem stuck. They were green before the re-indentation formatting fix.
